File: src/parser_utils/ssa_api.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

module type S = sig
  module L : Loc_sig.S

  type read_loc = L.t

  type write_loc =
    | Write of L.t Reason.virtual_reason
    | Uninitialized

  type write_locs = write_loc list

  type values = write_locs L.LMap.t

  val uninitialized : write_loc

  val write_locs_of_read_loc : values -> read_loc -> write_locs

  val is_dead_write_loc : values -> L.t -> bool
end

module Make (L : Loc_sig.S) : S with module L = L = struct
  module L = L

  type read_loc = L.t

  type write_loc =
    | Write of L.t Reason.virtual_reason
    | Uninitialized

  type write_locs = write_loc list

  type values = write_locs L.LMap.t

  let uninitialized = Uninitialized

  let write_locs_of_read_loc values read_loc = L.LMap.find read_loc values

  let is_dead_write_loc values loc =
    not
      (L.LMap.exists
         (fun _ write_locs ->
           Base.List.exists
             ~f:(function
               | Write r -> Reason.poly_loc_of_reason r = loc
               | Uninitialized -> false)
             write_locs)
         values)
end

module With_Loc = Make (Loc_sig.LocS)
module With_ALoc = Make (Loc_sig.ALocS)
include With_Loc

let print_values =
  let print_write_loc write_loc =
    match write_loc with
    | Uninitialized -> "(uninitialized)"
    | Write reason ->
      let loc = Reason.loc_of_reason reason in
      Utils_js.spf
        "%s: (%s)"
        (Loc.debug_to_string loc)
        Reason.(desc_of_reason reason |> string_of_desc)
  in
  fun values ->
    let kvlist = Loc_collections.LocMap.bindings values in
    let strlist =
      Base.List.map
        ~f:(fun (read_loc, write_locs) ->
          Printf.sprintf
            "%s => { %s }"
            (Loc.debug_to_string read_loc)
            (String.concat ", " @@ Base.List.map ~f:print_write_loc write_locs))
        kvlist
    in
    Printf.sprintf "[ %s ]" (String.concat "; " strlist)

File: rust_port/crates/flow_analysis/src/ssa_api.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::BTreeMap;
use std::hash::Hash;

use dupe::Dupe;
use flow_common::reason::VirtualReason;
use flow_common::reason::string_of_desc;
use flow_parser::loc::Loc;
use flow_parser::loc_sig::LocSig;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum WriteLoc<L: Dupe + Clone + Eq + Hash> {
    Write(VirtualReason<L>),
    Uninitialized,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Values<L: Dupe + Clone + Eq + Ord + Hash>(pub BTreeMap<L, Vec<WriteLoc<L>>>);

impl<L: Dupe + Clone + Eq + Ord + Hash> Values<L> {
    pub fn empty() -> Self {
        Values(BTreeMap::new())
    }

    pub fn contains_key(&self, key: &L) -> bool {
        self.0.contains_key(key)
    }

    pub fn get(&self, key: &L) -> Option<&Vec<WriteLoc<L>>> {
        self.0.get(key)
    }

    pub fn insert(&mut self, key: L, value: Vec<WriteLoc<L>>) {
        self.0.insert(key, value);
    }

    pub fn iter(&self) -> impl Iterator<Item = (&L, &Vec<WriteLoc<L>>)> {
        self.0.iter()
    }
}

pub fn write_locs_of_read_loc<L: Dupe + Clone + Eq + Ord + Hash>(
    values: &Values<L>,
    read_loc: &L,
) -> Vec<WriteLoc<L>> {
    values.0.get(read_loc).cloned().unwrap()
}

pub fn print_values(values: &Values<Loc>) -> String {
    fn print_write_loc(write_loc: &WriteLoc<Loc>) -> String {
        match write_loc {
            WriteLoc::Uninitialized => "(uninitialized)".to_string(),
            WriteLoc::Write(reason) => {
                let loc = reason.loc();
                format!(
                    "{}: ({})",
                    loc.debug_to_string(true),
                    string_of_desc(&reason.desc)
                )
            }
        }
    }

    let kvlist: Vec<_> = values.0.iter().collect();
    let strlist: Vec<String> = kvlist
        .iter()
        .map(|(read_loc, write_locs)| {
            let write_strs: Vec<String> = write_locs.iter().map(print_write_loc).collect();
            format!(
                "{} => {{ {} }}",
                read_loc.debug_to_string(true),
                write_strs.join(", ")
            )
        })
        .collect();
    format!("[ {} ]", strlist.join("; "))
}
